csharp/src/Apache.Arrow.Flight/Middleware/Interceptors/ClientInterceptorAdapter.cs

```diff
@@ -0,0 +1,161 @@
+            var callOptions = InterceptCall(context, out var middlewares);
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method, context.Host, callOptions);
+
+            var call = continuation(request, newContext);
+
+            var responseHeadersTask = call.ResponseHeadersAsync.ContinueWith(task =>
+            {
+                if (task.Exception == null && task.Result != null)
+                {
+                    var headers = task.Result;
+                    foreach (var m in middlewares)
+                        m?.OnHeadersReceived(new CallHeaders(headers));
+                }
+
+                return task.Result;
+            });
+
+            var wrappedResponseStream = new MiddlewareResponseStream<TResponse>(
+                call.ResponseStream,
+                call,
+                middlewares);
+
+            return new AsyncServerStreamingCall<TResponse>(
+                wrappedResponseStream,
+                responseHeadersTask,
+                call.GetStatus,
+                call.GetTrailers,
+                call.Dispose);
+        }
+
+
+        private CallOptions InterceptCall<TRequest, TResponse>(
+            ClientInterceptorContext<TRequest, TResponse> context,
+            out List<IFlightClientMiddleware> middlewareList)
+            where TRequest : class
+            where TResponse : class
+        {
+            var callInfo = new CallInfo(context.Method.FullName, context.Method.Type);
+
+            var headers = context.Options.Headers ?? new Metadata();
+            middlewareList = new List<IFlightClientMiddleware>();
+
+            var callHeaders = new CallHeaders(headers);
+
+            foreach (var factory in _factories)
+            {
+                var middleware = factory.OnCallStarted(callInfo);
+                middleware?.OnBeforeSendingHeaders(callHeaders);
+                middlewareList.Add(middleware);
+            }
+
+            return context.Options.WithHeaders(headers);
+        }
+
+        private async Task<TResponse> HandleResponse<TResponse>(
+            Task<TResponse> responseTask,
+            Task<Metadata> headersTask,
+            Func<Status> getStatus,
+            Func<Metadata> getTrailers,
+            Action dispose,
+            List<IFlightClientMiddleware> middlewares)
+        {
+            try
+            {
+                var headers = await headersTask.ConfigureAwait(false);
+                foreach (var m in middlewares)
+                {
+                    m?.OnHeadersReceived(new CallHeaders(headers));
+                }
+
+                var response = await responseTask.ConfigureAwait(false);
+                foreach (var m in middlewares)
+                {
+                    m?.OnCallCompleted(getStatus(), getTrailers());
+                }
+
+                return response;
+            }
+            catch
+            {
+                foreach (var m in middlewares)
```

Comment:
If the very last call to `OnCallCompleted` throws an exception, we're going to cycle through the list a second time calling all the middleware again. Consider instead catching errors inside of the enumeration and then only bubbling the exception out at the very end -- perhaps wrapping in an `AggregateException` if there is more than one.